feat(opentelemetry source): add support for enriching logs with HTTP headers

jblazquez opened the pull request:
## Summary

This PR adds a new `http.headers` configuration to the `opentelemetry` source, which injects HTTP headers received in the OTLP/HTTP request into the log events, similarly to how [`http_server`](https://vector.dev/docs/reference/configuration/sources/http_server/#headers) does it. Please see #21673 for the motivation behind this change.

This is my first contribution to Vector, so I'm not sure whether the changes in this PR are in the right form to be accepted. I wanted to create the PR to show other contributors the proposed implementation, but I'm open to different approaches.

## Change Type
- [ ] Bug fix
- [X] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [X] No

## How did you test this PR?

Added unit tests.

Also tested with an existing OTEL processing pipeline, replacing an existing `http_server` source (see below) with an `opentelemetry` source using the new `http.headers` config from this change.

Also ran a `wrk` load test to verify that performance is much higher with this solution than with an `http_source` that looks like this:

```
[sources.opentelemetry_logs_http_server]
type = "http_server"
address = "0.0.0.0:14318"
path = "/v1/logs"
decoding.codec = "protobuf"
decoding.protobuf.desc_file = "/path/to/opentelemetry-proto.desc"
decoding.protobuf.message_type = "opentelemetry.proto.collector.logs.v1.ExportLogsServiceRequest"
headers = [
    # To get access to the client IP address. 
    "x-forwarded-for",
]
```

See the linked issue (#21673) for more info on the performance impact of this new feature.

## Does this PR include user facing changes?

- [X] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist
- [X] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).

## References

Closes: #21673

jblazquez commented on src/sources/opentelemetry/http.rs:
```diff
@@ -103,12 +114,69 @@ pub(crate) fn build_warp_filter(
     log_filters.or(trace_filters).unify().boxed()
 }
 
+fn enrich_events(
```
This function is similar to [`SimpleHttpSource::enrich_events`](https://github.com/vectordotdev/vector/blob/master/src/sources/http_server.rs#L428), but it only injects headers into the event.

pront replied:
It looks like the code blocks that injects the headers are identical (except for the source names). I think it would be nice to convert this into a common function and move it to a new file `src/sources/util/http.rs`.

jblazquez commented on src/sources/opentelemetry/http.rs:
```diff
@@ -21,13 +22,21 @@ use vector_lib::tls::MaybeTlsIncomingStream;
+};
 
 use crate::http::{KeepaliveConfig, MaxConnectionAgeLayer};
+use crate::sources::http_server::{HttpConfigParamKind, SimpleHttpConfig};
```
Not entirely sure if it's OK to use types from the `source::http_server` module, but other sources (e.g. `heroku_logs`) also do it, so I assumed it is OK.

pront replied:
We can also move `HttpConfigParamKind` into the common file.

jblazquez replied:
I decided against making this change for now, as it touches a number of unrelated files.

jblazquez commented on src/sources/opentelemetry/tests.rs:
```diff
@@ -48,6 +48,7 @@ async fn receive_grpc_logs_vector_namespace() {
                 address: http_addr,
                 tls: Default::default(),
                 keepalive: Default::default(),
+                headers: Default::default(),
```
I tried adding a test for the new functionality, but none of the existing tests exercise the OTLP/HTTP endpoints, so it was not clear how I could do that.

pront replied:
I can see how this would add a lot of scope here. Since this change doesn't alter existing behavior, we can accept this contribution as is.

jblazquez replied:
I gave it another look and was able to figure out how to add the tests.

pront commented on src/sources/opentelemetry/http.rs:
```diff
@@ -103,12 +114,69 @@ pub(crate) fn build_warp_filter(
+                                        headers_config.get(header_name).map(HeaderValue::as_bytes);
+
+                                    log_namespace.insert_source_metadata(
+                                        SimpleHttpConfig::NAME,
```
Replace with `OpentelemetryConfig` (or this can also resolved by the refactoring I suggested above).

pront reviewed (commented):
Hi @jblazquez, thank you for this contribution!

jblazquez commented:
@pront, thank you for the review. Now that I'm more confident about these changes, I reworked the PR to address your feedback:

- 516d1daf63c3150f388a52ed1d682393cd11d6d3: Creates a new `sources::util::http::headers` module and adds an `add_headers`  function there. I tried to follow the existing pattern for the `sources::util::http::query::add_query_parameters` function, including the Cargo features.
- 36895e3c8a8e6fdf97b5daed6c70baa92356db4f: Modifies the `http_server` source to use the new `add_headers` function.
- 43f474f067821642d56f36f498c7c0d822e7fb41: Modifies the `opentelemetry` source to use the new `add_headers` function, adds tests, and updates the documentation and changelog.
- 0bf233c123818eb62cb25a5bcb4c9eed5f04fcf8: Corrects the description of the behavior of the `headers` option in `http_server`. Unlike `query_parameters`, the headers do not overwrite existing fields with the same name. Added a test to confirm the correct behavior.

In addition to the commits above, I also added the following commit that extends this PR to also bring support for the `host_key` configuration to the `opentelemetry` source, for more consistency with `http_server`. If you think this change should be done in a different PR please let me know, but I felt it was closely related to the `headers` feature.

- ea5c56ff4d87da0eddcac88e63ecde17d5570503: Adds support for the `host_key` config to `opentelemetry`. I followed the existing pattern from `http_server` for plumbing through the remote IP.

pront commented on src/sources/opentelemetry/http.rs:
```diff
@@ -59,6 +68,8 @@ pub(crate) async fn run_http_server(
 
     let span = Span::current();
     let make_svc = make_service_fn(move |conn: &MaybeTlsIncomingStream<TcpStream>| {
+        let remote_addr = conn.peer_addr();
+        let remote_addr_ref = enable_source_ip.then_some(remote_addr);
```
The following is preferable because it's `Some` only `enable_source_ip` is `true`:
```suggestion
 let remote_addr = enable_source_ip.then(|| conn.peer_addr());
```

jblazquez replied:
Thanks, if I come back to this work to add `host_key` to `opentelemetry` I will do this. I had copied what the `http_server` source did.

pront commented on changelog.d/21673_add_http_headers_to_opentelemetry_source.feature.md:
```diff
@@ -0,0 +1,3 @@
+The `opentelemetry` source can now be configured to enrich log events with the source IP and any HTTP headers received in the OTLP/HTTP request.
```
Let's also mention that now a `host_key` can be specified. 

Note, I am skeptical about the connection of the `host_key` with `source_ip`. I understand there's precedence in other sources but I can see an argument in favor of making it a different setting.

jblazquez replied:
Commit ea5c56ff4d87da0eddcac88e63ecde17d5570503 mentioned this, but I reverted that commit to reduce the scope of this PR, since you said you were not sure that `host_key` should mean `source_ip`.

pront commented on src/sources/opentelemetry/http.rs:
```diff
@@ -68,6 +79,13 @@ pub(crate) async fn run_http_server(
                 )
             }))
+            .map_request(move |mut request: hyper::Request<_>| {
+                if let Some(remote_addr_inner) = remote_addr_ref.as_ref() {
+                    request.extensions_mut().insert(*remote_addr_inner);
+                }
```
```suggestion
    if let Some(addr) = remote_addr {
        request.extensions_mut().insert(addr);
    }
```

pront commented on src/sources/opentelemetry/tests.rs:
```diff
@@ -334,6 +341,337 @@ async fn receive_grpc_logs_legacy_namespace() {
+            .unwrap()
+            .is_timestamp());
+        assert_eq!(
+            meta.get(path!("opentelemetry", "host")).unwrap(),
```
I would expect this to be `remote_ip`. I think this is also broken in `http_server` because we hardcode `path!("host")` when enriching events.

jblazquez replied:
I also wondered this, but I assumed it was intentional that - when using log namespacing - the source IP was always under `path!("host")` and that the `host_key` configuration was only for the legacy namespace.

But you bring up some good questions about the behavior of this setting, so I reverted the `host_key` change from this PR.

pront commented:
> @pront, thank you for the review. Now that I'm more confident about these changes, I reworked the PR to address your feedback:
> 
> * [516d1da](https://github.com/vectordotdev/vector/commit/516d1daf63c3150f388a52ed1d682393cd11d6d3): Creates a new `sources::util::http::headers` module and adds an `add_headers`  function there. I tried to follow the existing pattern for the `sources::util::http::query::add_query_parameters` function, including the Cargo features.
> * [36895e3](https://github.com/vectordotdev/vector/commit/36895e3c8a8e6fdf97b5daed6c70baa92356db4f): Modifies the `http_server` source to use the new `add_headers` function.
> * [43f474f](https://github.com/vectordotdev/vector/commit/43f474f067821642d56f36f498c7c0d822e7fb41): Modifies the `opentelemetry` source to use the new `add_headers` function, adds tests, and updates the documentation and changelog.
> * [0bf233c](https://github.com/vectordotdev/vector/commit/0bf233c123818eb62cb25a5bcb4c9eed5f04fcf8): Corrects the description of the behavior of the `headers` option in `http_server`. Unlike `query_parameters`, the headers do not overwrite existing fields with the same name. Added a test to confirm the correct behavior.
> 
> In addition to the commits above, I also added the following commit that extends this PR to also bring support for the `host_key` configuration to the `opentelemetry` source, for more consistency with `http_server`. If you think this change should be done in a different PR please let me know, but I felt it was closely related to the `headers` feature.
> 
> * [ea5c56f](https://github.com/vectordotdev/vector/commit/ea5c56ff4d87da0eddcac88e63ecde17d5570503): Adds support for the `host_key` config to `opentelemetry`. I followed the existing pattern from `http_server` for plumbing through the remote IP.

Thank you @jblazquez, much appreciated 👍 

Adding the `host_key` introduced new scope so I did another review and some new things came up.

jblazquez commented:
@pront Appreciate the review!

You brought up some good questions about how `host_key` should behave, so I did not want to add that new scope to this PR. I reverted the commit that added `host_key` so this PR is purely about the `headers` config.

Let me know if there are other changes needed. Thanks!

pront commented:
> @pront Appreciate the review!
> 
> You brought up some good questions about how `host_key` should behave, so I did not want to add that new scope to this PR. I reverted the commit that added `host_key` so this PR is purely about the `headers` config.
> 
> Let me know if there are other changes needed. Thanks!

Sounds great. Will review again today.

If you are still interested in adding a `host_key`, feel free to submit another after this one is merged. I think we have a good understanding of how it should work.

pront reviewed (approved):
Thank you @jblazquez!
